python/lsst/pipe/tasks/imageDifference.py

```diff
@@ -367,23 +368,175 @@ def runDataRef(self, sensorRef, templateIdList=None):
         expId = int(sensorRef.get("ccdExposureId"))
         idFactory = afwTable.IdFactory.makeSource(expId, 64 - expBits)
 
+        if self.config.doAddCalexpBackground:
+            calexpBackgroundExposure = sensorRef.get("calexpBackground")
+
         # Retrieve the science image we wish to analyze
         exposure = sensorRef.get("calexp", immediate=True)
+
+        # Retrieve the template image
+        template = self.getTemplate.run(exposure, sensorRef, templateIdList=templateIdList)
+        templateExposure = template.exposure
+        templateSources = template.sources
+
+        # Load the calexp source catalog or perform detection anyway
+        # it can be used in the doDetection branch in run()
+        if not sensorRef.datasetExists("src"):
+            sciencePsf = exposure.getPsf()
+            scienceSigmaOrig = sciencePsf.computeShape().getDeterminantRadius()
+            if self.config.doPreConvolve:
+                scienceSigmaPost = scienceSigmaOrig*math.sqrt(2)
+            else:
+                scienceSigmaPost = scienceSigmaOrig
+            self.log.warn("Src product does not exist; running detection, measurement, selection")
+            # Run own detection and measurement; necessary in nightly processing
+            selectSources = self.subtract.getSelectSources(
+                exposure,
+                sigma=scienceSigmaPost,
+                doSmooth=not self.doPreConvolve,
+                idFactory=idFactory,
+            )
+        else:
+            self.log.info("Source selection via src product")
+            # Sources already exist; for data release processing
+            selectSources = sensorRef.get("src")
+
+        # Prepare kernelSources for AL either based on the template or on
+        # external reference catalog
+        if self.config.doSubtract:
+            if self.config.subtract.name == 'al':
+                if self.config.doSelectSources:
```

Comment:
This may be beyond the scope of this ticket, but I would really like to see the code in the `if self.config.doSelectSources:` block moved to its own method. This would include the code in the `if self.config.doSelectSources:` block at line 578 (I'll make a similar comment there).

Reply:
Added  todo comments.